authorize_helper.go

```diff
@@ -172,10 +168,10 @@ func IsValidRedirectURI(redirectURI *url.URL) bool {
 }
 
 func IsRedirectURISecure(redirectURI *url.URL) bool {
-	return !(redirectURI.Scheme == "http" && !IsLocalhost(redirectURI))
+	return redirectURI.Scheme == "https" || (redirectURI.Scheme == "http" && IsLocalhost(redirectURI))
```

Comment:
This is what the spec says is secure, not sure why would any other scheme be secure? Other schemes are not secure. So I moved code from `OpenIDConnectRequestValidator` here, and made `OpenIDConnectRequestValidator` use `IsRedirectURISecure`.

Reply:
> Secure in this context does not mean "HTTPS" or "encrypted".

Yes, I mean really secure. When I wrote above, "This is what the spec says is secure, not sure why would any other scheme be secure" I meant [RFC 8252](https://tools.ietf.org/html/rfc8252). And I think you are referring to the same one. We just disagree in how we read it. :-) But I even disagree with the spec and I think some things they do not forbid explicitly are insecure.

In any case, I can change the default here because now it is configurable. But I do think it is better to be secure out of the box and if somebody wants to relax security, they can configure a less secure version of "IsRedirectURISecure". I think this is what fosite is all about. I think fosite should see as secure only https (and claimed https redirects) and loopback redirects.

> custom-scheme URLs for mobile apps are loopback, they are considered secure

No, they are not. See [here](https://developer.apple.com/documentation/xcode/allowing_apps_and_websites_to_link_to_your_content/defining_a_custom_url_scheme_for_your_app):

> URL schemes offer a potential attack vector into your app, so make sure to validate all URL parameters and discard any malformed URLs. In addition, limit the available actions to those that don’t risk the user’s data. For example, don’t allow other apps to directly delete content or access sensitive information about the user. When testing your URL-handling code, make sure your test cases include improperly formatted URLs.

If a company who offers custom schemes writes this, I think we should believe them, no?

It is true that the spec [in appendix B.1](https://tools.ietf.org/html/rfc8252#appendix-B.1) says:

> To receive the authorization response, both private-use URI scheme (referred to as "custom URL scheme") redirects and claimed "https" scheme URIs (known as "Universal Links") are viable choices.

But also:

> Claimed "https" scheme URIs are the preferred redirect choice on iOS 9 and above due to the ownership proof that is provided by the operating system.

In [section 8.1](https://tools.ietf.org/html/rfc8252#section-8.1) the spec acknowledges this:

> A limitation of using private-use URI schemes for redirect URIs is that multiple apps can typically register the same scheme, which makes it indeterminate as to which app will receive the authorization code.

And they offer PKCE as a solution to interception. And where I disagree with the spec. PKCE only prevents a malicious app to intercept OIDC request originating from true app. But it **does not** protect against malicious app **both initiating sign-in flow and then receiving the redirect**. In fact, true app might not even be installed on the phone, just a malicious app might be. Which means that it could phish an user to sign-in with my fosite-based authentication service, obtaining an access token they can then misuse.

So given that native apps today have access to "claimed https redirects" on both iOS and Android platforms and that this is secure, why allow anything less? Why open a potential window of attack, given that a more secure alternative exists?

Reply:
> So given that native apps today have access to "claimed https redirects" on both iOS and Android platforms and that this is secure, why allow anything less? Why open a potential window of attack, given that a more secure alternative exists?

Because not doing that will cause breaking changes for existing apps, which in turn means potentially hundreds of people complaining that ORY Hydra no longer works. It is very, very common to use custom redirect URIs in mobile environments.

Your interpretation might be correct or better, but the real world forces our hand here. Because the specification is simply the way it is, it is implemented as such very commonly in the real world. So let's not make our user's life harder than it has to be, OAuth2 sucks enough already.